python/src/scipp/coords.py

```diff
@@ -0,0 +1,268 @@
+    return obj.coords[name]
+
+
+class CoordTransform:
+    Graph = Dict[Union[str, Tuple[str, ...]], Union[str, Callable]]
+
+    def __init__(self, obj, *, graph, outputs):
+        self.obj = obj.copy(deep=False)
+        # TODO We manually shallow-copy the buffer, until we have a better
+        # solution for how shallow copies also shallow-copy event buffers.
+        if self.obj.bins is not None:
+            self.obj.data = bins(**self.obj.bins.constituents)
+        self._original = obj
+        self._events_copied = False
+        self._rename = {}
+        self._memo = []  # names of product for cycle detection
+        self._aliases = []  # names that alias other names
+        self._consumed = []  # names that have been consumed
+        self._outputs = outputs  # names of outputs
+        self._graph = graph
+
+    def _add_event_coord(self, key, coord):
+        try:
+            self.obj.bins.coords[key] = coord
+        except VariableError:  # Thrown on mismatching bin indices, e.g. slice
+            self.obj.data = self.obj.data.copy()
+            self.obj.bins.coords[key] = coord
+
+    def _add_event_coords(self, coords):
+        for key, coord in coords.items():
+            # Non-binned coord should be duplicate of dense handling above,
+            # if present => ignored.
+            if coord.bins is not None:
+                self._add_event_coord(key, coord)
+
+    def _add_coord(self, *, name):
+        if name in self.obj.meta:
+            return _produce_coord(self.obj, name)
+        if isinstance(self._graph[name], str):
+            self._aliases.append(name)
+            out = self._get_coord(self._graph[name])
+            if self.obj.bins is not None:
+                # Calls to _get_coord for dense coord handling take care of
+                # recursion and add also event coords, here and below we thus
+                # simply consume the event coord.
+                if self._graph[name] in self.obj.meta:
+                    out_bins = _consume_coord(self.obj.bins, self._graph[name])
+            dim = (self._graph[name], )
+        else:
+            func = self._graph[name]
+            argnames = _argnames(func)
+            args = {arg: self._get_coord(arg) for arg in argnames}
+            out = func(**args)
+            if self.obj.bins is not None:
+                args.update({
+                    arg: _consume_coord(self.obj.bins, arg)
+                    for arg in argnames if arg in self.obj.bins.meta
+                })
+                out_bins = func(**args)
+            dim = tuple(argnames)
+        if isinstance(out, Variable):
+            out = {name: out}
+        self._rename.setdefault(dim, []).extend(out.keys())
+        for key, coord in out.items():
+            self.obj.coords[key] = coord
+        if self.obj.bins is not None:
+            if isinstance(out_bins, Variable):
+                out_bins = {name: out_bins}
+            self._add_event_coords(out_bins)
```

Comment:
I think these transformations require a 1:1 input to output. For cases where you might lose/gain extent, such as in wfm stitching, would that be possible at present?

Reply:
I don't understand what you are referring to, can you explain?

Reply:
More a conceptual question of what we can do with this rather than a problem I can see per-say. Going back to our TOF -> Wavelength example, for every tof bin we have a corresponding wavelength bin we can trivially create a function to handle this transformation and reference in our graph. However Neil and I have been talking about using this mechanism for WFM, but in order to do that we would need to take sub-sections of the input coordinate (t), convert to TOF and stitch together. I guess the main concern is that the converted coordinate uniquely references the dimension... I may just need to try this out.

Reply:
I think these are just two separate steps:
1. `transform_coords`
2. Stitch.